### DISCUSSION TITLE
SELECT on parent performed when committing DELETE on child even without cascading nor mappings to update

### DISCUSSION THREAD

**Discussion by lesinigo** [sqlalchemy/sqlalchemy#10596] [Usage Questions] [answered]
### Describe the bug

With a many-to-one lazy loaded relationship from Child to Parent, without an actual SQL foreign key, SQLAlchemy is performing a SELECT on the Parent when deleting the Child even if there is no cascade nor any Parent field to update.

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.lazy

### SQLAlchemy Version in Use

2.0.23

### DBAPI (i.e. the database driver)

pysqlite

### Database Vendor and Major Version

SQLite

### Python Version

3.11.6, 3.12.0

### Operating system

Linux, OSX

### To Reproduce

```python
#!/usr/bin/env python3

from sqlalchemy import create_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, Session, foreign, mapped_column, relationship, remote

class Base(DeclarativeBase):
    """SQLAlchemy declarative base"""

class Parent(Base):
    __tablename__ = "parent"
    id: Mapped[int] = mapped_column(primary_key=True)

class Child(Base):  # pylint:disable=too-few-public-methods
    __tablename__ = "child"

    name: Mapped[str] = mapped_column(primary_key=True)
    parent_id: Mapped[int] = mapped_column()

    # custom foreign condition without SQL-level foreign keys
    parent: Mapped[Parent] = relationship(
        primaryjoin=remote(Parent.id) == foreign(parent_id),
        cascade="none",
        lazy="select"
    )

engine = create_engine("sqlite:///:memory:", echo=False)
Base.metadata.create_all(engine)
session = Session(bind=engine)
# NB: we are not creating any Parent
child = Child(name="some_child", parent_id=0)
session.add(child)
session.commit()
engine.echo = True

# test case
print("******************************** delete")
session.delete(child)
print("******************************** commit after delete")
session.commit()
# here SQLAlchemy will perform a SELECT FROM parent before a DELETE FROM child
```

### Error

```
******************************** delete
******************************** commit after delete
2023-11-06 21:14:04,374 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-11-06 21:14:04,386 INFO sqlalchemy.engine.Engine SELECT child.name AS child_name, child.parent_id AS child_parent_id 
FROM child 
WHERE child.name = ?
2023-11-06 21:14:04,394 INFO sqlalchemy.engine.Engine [generated in 0.00887s] ('some_child',)
2023-11-06 21:14:04,400 INFO sqlalchemy.engine.Engine SELECT parent.id AS parent_id 
FROM parent 
WHERE parent.id = ?
2023-11-06 21:14:04,405 INFO sqlalchemy.engine.Engine [generated in 0.00505s] (0,)
2023-11-06 21:14:04,408 INFO sqlalchemy.engine.Engine DELETE FROM child WHERE child.name = ?
2023-11-06 21:14:04,414 INFO sqlalchemy.engine.Engine [generated in 0.00614s] ('some_child',)
2023-11-06 21:14:04,416 INFO sqlalchemy.engine.Engine COMMIT
``` 

### Additional context

Documentation about `lazy ` says:

> specifies How the related items should be loaded. Default value is select. Values include:
> select - items should be loaded lazily when the property is first accessed, using a separate SELECT statement, or identity map fetch for simple many-to-one references.

And documentation about `passive_deletes` says:
> Normally, when a parent item is deleted, all child items are loaded so that they can either be marked as deleted, or have their foreign key to the parent set to NULL. 

Here we have `cascade="none"` and the default value for `lazy="select"`, there is no mapping of Child collection on the Parent, also there is no `Parent` instance in the identity map - there's actually no Parent at all both in the session and in the database.

It is my understanding that SQLAlchemy should not perform a `SELECT FROM parent` in this scenario because:

- nothing is accessing `child.parent`
- there is no foreign key or other relation from Parent to Child to be updated
- we are not cascading anything
- we are actually deleting a child, not the parent (ie. we are deleting the "many" side of the many-to-one)

I am not using an actual SQL ForeignKey because in my use case it is expected and normal to have childs referring to parent IDs that do not exist in the parent table.

Tested variations:

- not explicitly setting `cascade` does not change anything
- setting `cascade="all"` causes the `SELECT FROM parent` to be performed after `session.delete(child)` instead of after `session.commit()`, still unexpected to me for the reasons listed above
- setting `passive_deletes=True` on the `Child.parent` relationship makes the unexpected `SELECT FROM parent` to disappear, which is ok but AFAIK should not be needed

The above test case has been reproduced in clean environments on Linux with Python 3.11.6, on macOS with Python 3.11.6 and on macOS with Python 3.12.0, always using the latest stable release of SQLAlchemy 2.0.23 installed through pip.

**Comment by zzzeek** [comment]
it's not clear there is any "bug" here, though the documentation shows the form you are using:

```
    parent: Mapped[Parent] = relationship(
        primaryjoin=remote(Parent.id) == foreign(parent_id),
        cascade="none",
        lazy="select"
    )
```

this form is insufficient at the moment because "parent_id" has not yet been mapped, which is leading the "use get" condition to revert to False:

```
>>> print(Child.__mapper__.attrs.parent.strategy.use_get)
False
```


using a lambda allows the `foreign()` annotation to only be called once the `parent_id` column is fully initialized:

```py
    parent: Mapped[Parent] = relationship(
        primaryjoin=lambda: remote(Parent.id) == foreign(Child.parent_id),
        cascade="none",
        lazy="select"
    )
```

will look into the annotation thing.

**Reply by zzzeek** [reply]
yah that parent_id column has no type yet.  applying foreign() to it makes a copy of it, which has no type.  the `Integer` type being set is not propagated into the `primaryjoin` and I dont at the moment see any way it would be possible.  

so far I think the most reasonable course of action here is to emit a warning / raise an error if foreign() or remote() are applied to columns not set up with a type yet

**Reply by zzzeek** [reply]
ill have a fix in #10597

**Reply by lesinigo** [reply]
yeah I'm not 100% sure that it would strictly be classified as a "bug", at least not in the sense that something's broken, but from my expectations as a user after reading docs I did not expected an extra query. I guess it could have a non-trivial db load / performance impact on some DELETE-intensive workloads involving models configured as the provided example.

I can confirm that the lambda approach solves the extra query and it works even with all relationship settings at their default (ie. not specifying `cascade` nor `lazy`), I've already adopted it for this use case:

```python
class Child(Base):
    __tablename__ = "child"
    name: Mapped[str] = mapped_column(primary_key=True)
    parent_id: Mapped[int] = mapped_column()

    # custom foreign condition without SQL-level foreign keys
   parent: Mapped[Parent] = relationship(primaryjoin=lambda: remote(Parent.id) == foreign(Child.parent_id))
```

**Comment by lesinigo** [comment]
As far as I can tell #10597 was included in [SQLAlchemy 2.0.24](https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-6c7bfe85d6e00ea3547f85d8a6759e94) but I'm still getting the extra SELECT described above even with the latest SA 2.0.29.

I've also found a different use case that IMHO is more common and can have a bigger impact on workloads: one extra SELECT gets issued for every single row that gets deleted from an adjacency list. It is also a simpler code to reproduce.

### SQLAlchemy Version in Use

2.0.29

### DBAPI (i.e. the database driver)

pysqlite

### Database Vendor and Major Version

SQLite

### Python Version

3.11.8, 3.12.3

### Operating system

Linux, OSX

### To Reproduce

```python
#!/usr/bin/env python3

"""SQLAlchemy test to reproduce weird SELECT-before-DELETE case"""

from sqlalchemy import create_engine, ForeignKey, select
from sqlalchemy.orm import DeclarativeBase, Mapped, Session, mapped_column, relationship


class Base(DeclarativeBase):
    """SQLAlchemy declarative base"""


class Node(Base):
    __tablename__ = "nodes"
    id: Mapped[int] = mapped_column(primary_key=True)
    parent_id: Mapped[int | None] = mapped_column(ForeignKey("nodes.id"))
    parent: "Mapped[Node | None]" = relationship(back_populates="children", remote_side="Node.id")
    children: "Mapped[list[Node]]" = relationship(back_populates="parent")


engine = create_engine("sqlite:///:memory:", echo=True)
Base.metadata.create_all(engine)
session = Session(bind=engine)
session.add(Node(id=999))  # root node
session.add_all([Node(id=_, parent_id=999) for _ in range(100)])
session.commit()
session.expunge_all()

# test case
print("******************************** delete")
childs = session.execute(select(Node).where(Node.parent_id == 999)).scalars().all()
for child in childs:
    session.delete(child)
print("******************************** commit after delete")
session.commit()
# here SQLAlchemy will perform 100 selects (one per deleted row) before issuing a single delete for all deleted PKs
```

**Reply by zzzeek** [reply, marked as answer]
I ran the original test case at the top, the output is:

```
******************************** delete
******************************** commit after delete
2024-04-18 09:28:08,515 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-04-18 09:28:08,516 INFO sqlalchemy.engine.Engine SELECT child.name AS child_name, child.parent_id AS child_parent_id 
FROM child 
WHERE child.name = ?
2024-04-18 09:28:08,517 INFO sqlalchemy.engine.Engine [generated in 0.00008s] ('some_child',)
2024-04-18 09:28:08,517 INFO sqlalchemy.engine.Engine DELETE FROM child WHERE child.name = ?
2024-04-18 09:28:08,517 INFO sqlalchemy.engine.Engine [generated in 0.00006s] ('some_child',)
2024-04-18 09:28:08,517 INFO sqlalchemy.engine.Engine COMMIT

```

there is no SELECT from parent

your second test is totally different.  the Node.children relationship is loading on each Node because it's attempting to check for elements that should be DELETEd.

Solution, add passive_deletes, which will assume other means are being used to emit the DELETE, I would recommend putting ON DELETE CASCADE on the foreign key constraint: https://docs.sqlalchemy.org/en/20/orm/cascades.html#passive-deletes

```
class Node(Base):
    __tablename__ = "nodes"
    id: Mapped[int] = mapped_column(primary_key=True)
    parent_id: Mapped[int | None] = mapped_column(ForeignKey("nodes.id"))
    parent: "Mapped[Node | None]" = relationship(back_populates="children", remote_side="Node.id")
    children: "Mapped[list[Node]]" = relationship(back_populates="parent", passive_deletes=True)

```

output:

```
******************************** delete
2024-04-18 09:32:58,522 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-04-18 09:32:58,523 INFO sqlalchemy.engine.Engine SELECT nodes.id, nodes.parent_id 
FROM nodes 
WHERE nodes.parent_id = ?
2024-04-18 09:32:58,523 INFO sqlalchemy.engine.Engine [generated in 0.00007s] (999,)
******************************** commit after delete
2024-04-18 09:32:58,525 INFO sqlalchemy.engine.Engine DELETE FROM nodes WHERE nodes.id = ?
2024-04-18 09:32:58,525 INFO sqlalchemy.engine.Engine [generated in 0.00022s] [(0,), (1,), (2,), (3,), (4,), (5,), (6,), (7,)  ... displaying 10 of 100 total bound parameter sets ...  (98,), (99,)]
2024-04-18 09:32:58,525 INFO sqlalchemy.engine.Engine COMMIT

```

**Reply by lesinigo** [reply]
thanks for the feedback! I'll need some time to go trough everything :)

**Reply by lesinigo** [reply]
I had time to re-check this issue more carefully and I can confirm that I'm no more getting SELECT-before-DELETE from my first test case, I don't know what I was looking at when I said I had it with SA 2.0.29, sorry.

I was also writing a reply about the second use case and having to go through that made me understand (I think) the reason of SELECTs in that use case. SQLAlchemy is basically doing `SELECT ... WHERE parent_id = <id of the one being deleted>` just to be able to eventually execute `UPDATE ... SET parent_id = None WHERE ... <big parameter set with individual PKs from previous query>` instead of collapsing the two into a much simpler/quicker `UPDATE table SET parent_id = NULL WHERE parent_id = <id of the one being deleted` and I suspect the reason is that in my use case none of the rows being updated is in SQLAlchemy identity map (the quicker single query would have zero side effects) but it would be hard / slow / worse in some way to handle generic use cases where part of them could be in the identity map possibly through complex relationships...
